Canonicalize accelerator names in Resources

WoosukKwon opened the pull request:
Fixes #995.

To prevent potential bugs due to lower-case names, I think we should canonicalize the accelerator names in `Resources`. Because there are only < 20 accelerators (except TPUs) in three clouds, we can maintain a hard-coded list of accelerator names. Using them, we can avoid reading the whole catalogs in most cases.

Note that this list (`_ACCELERATORS`) must also include accelerators supported for local clusters.

Tested:
* `sky exec --gpus v100`, `sky exec --gpus V100`
* `sky exec --gpus a10g`, `sky exec --gpus A10g`, `sky exec --gpus A10G`

concretevitamin commented on sky/utils/resources_utils.py:
```diff
@@ -0,0 +1,63 @@
+"""Resources related utilities."""
```
We can call this module accelerator_registry? So it reminds people if we want to add a new accelerator, this is the module to change (see a related comment below).

WoosukKwon replied:
Great point. Fixed!

concretevitamin commented on sky/utils/resources_utils.py:
```diff
@@ -0,0 +1,63 @@
+from sky.utils import ux_utils
+
+# List of all accelerators supported by SkyPilot.
+# NOTE: Must include accelerators supported for local clusters.
```
Can you document what happens if:
- a name is in this list, but not in any catalog (can users launch it?)
- a name is not in this list, but in a catalog (can users launch it?)
- for developers, what actions, if any, should we take to keep this list and the catalogs in sync?

WoosukKwon replied:
Added more comments. PTAL.

concretevitamin commented on sky/utils/resources_utils.py:
```diff
@@ -0,0 +1,63 @@
+    'T4g',
+    'V100',
+    'V100-32GB',
+    'Virtex UltraScale (VU9P)',
```
For example, `show-gpus --all` doesn't show this name or L11, possibly others.

concretevitamin commented on sky/utils/resources_utils.py:
```diff
@@ -0,0 +1,63 @@
+
+# List of all accelerators supported by SkyPilot.
+# NOTE: Must include accelerators supported for local clusters.
+_ACCELERATORS = [
```
Comment that these are the *canonicalized* accelerator names used internally. Users should call `canonicalize_accelerator_name()` to get them.

WoosukKwon replied:
Fixed.

concretevitamin commented on sky/utils/resources_utils.py:
```diff
@@ -0,0 +1,63 @@
+
+    if len(names) == 0:
+        with ux_utils.print_exception_no_traceback():
+            raise ValueError(f'Unknown accelerator: {accelerator}')
```
UX: 
```
(py37) ~ » sky launch --gpus V1000 ''                                                              2 ↵
I 08-14 08:29:48 optimizer.py:865] No resource satisfying {'V1000': 1} on [AWS, Azure, GCP].
sky.exceptions.ResourcesUnavailableError: No launchable resource found for task sky-cmd. To fix: relax its resource requirements.
Hint: 'sky show-gpus --all' to list available accelerators.
      'sky check' to check the enabled clouds.

(py37) ~ » sky launch --gpus V1000 ''                                                              1 ↵
ValueError: Unknown accelerator: V1000
```
Top is master branch's error message, bottom is this branch.  I feel like the former message is more informative. Can we fall back to that somehow, or move that to here?

WoosukKwon replied:
Good catch. The error will be handled by the optimizer.

concretevitamin commented on sky/utils/resources_utils.py:
```diff
@@ -0,0 +1,63 @@
+        return names[0]
+
+    if len(names) > 1:
+        with ux_utils.print_exception_no_traceback():
```
Comment on an example that could trigger this?

WoosukKwon replied:
Added a comment. This is unreachable for now.

concretevitamin commented on sky/utils/resources_utils.py:
```diff
@@ -0,0 +1,63 @@
+    if accelerator.lower() in mapping:
+        return mapping[accelerator.lower()]
+
+    # A new accelerator can be added after upgrading SkyPilot.
```
Is this meant to catch that  a user 
- just upgraded catalogs
- hasn't upgraded SkyPilot code so the hard-coded list above is still outdated?

WoosukKwon replied:
Yes. I clarified the comment. PTAL.

concretevitamin reviewed (commented):
This is awesome @WoosukKwon! Left some comments.

WoosukKwon commented:
Thanks for quick review @concretevitamin! Addressed all comments. PTAL.

BTW, I realized there is a potential problem in this approach. While this is NOT the case for our current catalogs, the clouds may add the same accelerator with different names (e.g., A10g and A10G) in the future. In such a case, its canonical name can be ambiguous. WDYT @concretevitamin? Should we also handle such cases?

concretevitamin commented on sky/utils/accelerator_registry.py:
```diff
@@ -1,9 +1,21 @@
-# List of all accelerators supported by SkyPilot.
+# Canonicalized names of all accelerators (except TPUs) supported by SkyPilot.
 # NOTE: Must include accelerators supported for local clusters.
+
```
styleguide nit: L7 should have a `#`, and L18 should not be blank --  to form a block of comments from L5 to L17 for L19.

WoosukKwon replied:
Thanks for letting me know! Fixed.

concretevitamin commented on sky/utils/accelerator_registry.py:
```diff
@@ -1,9 +1,21 @@
+# Optimizer will print an error message.
+# 2. What if a name is not in this list, but in a catalog?
+# The name will be searched on the catalog with its case being ignored.
+# If a match is found, the name will be canonicalized to that in the catalog.
```
In other words, is this list simply an optimization to short-circuit checking the catalog?  If so, should document here or point 3.

WoosukKwon replied:
Yes it is. Fixed the comment.

concretevitamin commented:
> BTW, I realized there is a potential problem in this approach. While this is NOT the case for our current catalogs, the clouds may add the same accelerator with different names (e.g., A10g and A10G) in the future. In such a case, its canonical name can be ambiguous. WDYT @concretevitamin? Should we also handle such cases?

If this happens, we could define two such canonical names and require any user input to match them, possibly banning case insensitivity (which you have already done in the "unreachable case").   I doubt hardware makers would choose an ambiguous name, however.

concretevitamin reviewed (approved):
LGTM.  Reminder of this comment: https://github.com/skypilot-org/skypilot/pull/1075#discussion_r945306264

WoosukKwon commented:
I worked well on local clusters as well. @concretevitamin Can I merge this?

concretevitamin commented:
LGTM - let's ship this

On Wed, Aug 17, 2022 at 1:36 PM Woosuk Kwon ***@***.***>
wrote:

> I worked well on local clusters as well. @concretevitamin
> <https://github.com/concretevitamin> Can I merge this?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/skypilot-org/skypilot/pull/1075#issuecomment-1218465963>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEQWHS44PW6WFIOAR42A33VZVELRANCNFSM56PP76UA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

WoosukKwon commented:
@istoica Could you add more? When do you think we might want to allow different names for the same accelerator?

WoosukKwon commented:
> BTW, I realized there is a potential problem in this approach. While this
 is NOT the case for our current catalogs, the clouds may add the same
 accelerator with different names (e.g., A10g and A10G) in the future. In
 such a case, its canonical name can be ambiguous. WDYT @concretevitamin
 <https://github.com/concretevitamin>? Should we also handle such cases?

Oh now I understand the context. Actually, I later realized that my comment was wrong. We have another layer that converts our internal representation of the accelerators (i.e., the canonicalized names) to the accelerator names defined by each cloud. Thus, SkyPilot can handle the case that clouds have different names for the same accelerator.
